### ISSUE TITLE
Protractor's functionality is inconsistent, slow, and not practical

### ISSUE THREAD

**Issue by Inframatic** [angular/protractor#4851]
Protractor will have no problem finding an element on one run and then with the same code on another run, fail, citing that it cannot find the element. Unfortunately this inconsistent behavior happens consistent for all different types tests.

I move that Protractor be moved back into beta as it is definitely not ready to be released in any sort of professional environment.

**Comment by kylecordes** [comment]
I think this is too harsh; many of us have been getting lots of practical value from Protractor tests for years. It is very useful, and has been for a long time. However, I can make some suggestions along the same lines as the original comment. I offer this in the spirit of helpful ideas, not complaining - I have already gotten so much value from Protractor.

Printing out the test name only after execution is inherited from Jasmine, where it is generally pretty harmless for the kind of testing done with Jasmine. But this is indeed a very frustrating developer ergonomic problem for users of Protractor. When it is stuck, you never know why it is thought, unless you go study your test suite and find out what the next thing is that it is working on. I don't think a protractor should reinvent all of Jasmine, but maybe patch in a hook? Lobby those authors to provide a mode that prints the test name first?

The documentation,  in combination with some searches of the issue tracker, generally can get past any specific practical problem. But it doesn't offer much advice on how to use Protractor in the most streamlined, efficient, ergonomic matter. Documentation that guides users down a path of the best possible use of the tools, even with no changes to the tools, could improve the perceived difficulty of Protractor.

The documentation mentions the "page objects" pattern, but doesn't (on my last read) mention that often page objects don't have any state in them - and therefore can be repalced with the even simpler "file with some functions and it about a specific page" idea, which trims lines of code and complexity.

Along the same lines as the documentation, it would be wonderful if the example code were significantly larger, more "real world", and dealt with timing issues of the kinds that cause developer frustration.

In terms of startup speed and the like, much of this is inherited from Selenium; although there isn't much Protractor can do about it, maybe a little bit of progress reporting, so that users are aware of what time was spent by what tool, would improve perceptions of Protractor.

In my experience, numerous projects ultimately require disabling the "wait for angular" feature; this happens when such tools depend on underlying libraries which themselves cause Protractor to believe it should keep waiting. I wonder if some documentation, or even a black list/white list included in the software, could detect dependencies on tools known to cause those difficulties. If users see upfront "I am having a more difficult experience with E2E testing because library X needlessly sets timeouts and intervals everywhere", they not see this as a Protractor problem per se.

In the presence of such libraries though, and accepting that "wait for angular" sometimes must be turned off, it would be ideal if the examples work through, and documentation included, a detailed treatment on how to have ergonomic testing, which very rarely fails because of random timing, in the absence of this feature. (For example, at the start of each bit of testing, use "browser.wait" expectations by which the test can confirm that software under test is in the starting state that the test author planned.)

(Over here we occassionaly teach classes on Protractor, hence have had the opportunity to create and update our own examples, talk about some of these things aren't addressed in the documentation very much, and so on.)

**Comment by qiyigg** [comment]
@kylecordes Thanks for great explanation and support.
@Inframatic Sorry to hear the bad experience you had with Protractor; however, I had to say, Protractor **is** practical and we do have lots of external users and internal users who are using it extensively. But yes, there are still lots of stuff that Protractor could do better, especially the document part. We are trying to make it better gradually, but slowly. It would really welcome if you and other users could contribute together to make it even better, since you are the best people who know where Protractor didn't do well and should get improved.

For the issues you mentioned. None of them seem normal except the last one, it would be better to provide some code examples to see what really happens. I guess other folks must encounter the similar issue you have.

From my experience:

Protractor will have no problem finding an element on one run and then with the same code on another run, fail, citing that it cannot find the element. This happens often.
If the page element does exist for each run. It's very likely the Protractor doesn't wait for the page to be stable properly, this could happen if you disable waitForAngular / miss "await" keyword for async webdriver call. 

The software takes 2 minutes to start and for many tasks can be up to 5x slower than doing that same task by hand.
Protractor shouldn't be significant slow than a regular webdriver test. If it is super slow, it is very likely Protractor is waiting for the page to be stable and the page has some outstanding async tasks that runs slow or have a long running delay. e.g. a setTimeOut(xx, 1000), Protractor will wait for it to be completed and then continue the test, since it might modify the page. If this kind of async task does change the page, we should wait for it although it might take long time; otherwise, we can either let this task run outside Angular zone or **temporarily** disable waitForAngular. 

Many times, it randomly times out in areas where it hadnt before for aspects of a project that do work when e2e tested by hand.
Usually caused by the same issue mentioned above.

For the waiting strategy, some what like  "black list/white list" mentioned by  @kylecordes , one of our folks is working on it. But he doesn't seem have enough time to check it in and also blocked by some potential Angular changes. The waitForAngular issue should be significantly eased after his change though.

**Comment by JunichiOzawa** [comment]
Hello, Just wanted to share my experience.

I am writing tests for my ionic3 + Angular5 application and I was using protractor 5.4.0(w/ jasmine 3.1.0) and all my tests were unstable  just like @Inframatic has explained above. They sometimes succeeded and sometimes failed(with index out of bounds, element not found or something of that sort). I tried all sorts of options in protractor.conf.js and technics mentioned here and there, like putting browser.sleeps, browser.waits and browser.waitForAngularEnabled(false). funny thing was, it went 90% stable when I run a single test with --specs option, but it was always unstable with multiple tests simultaneously.

Today, I modified the protractor version to 5.4.1 and jasmine to 2.9.1. I downgraded jasmine because the protractor official tutorial mentions about jasmine version 2.4 so I just gave it a try with version 2 series. Then... all my tests started working fine every time I run them. with or without the specs option.

Hope this information might help someone.

**Comment by demisx** [comment]
We've been running protractor 5.4.x tests with jasmine 3.4.x and Chrome web driver 2.46. We use direct connect on Chrome browser only. Stability has been notoriously a big issue for us - e.g. constant false failures with `element not interactable` type of errors, failing in headless vs passing in visual mode, failing when running as part of a suite, but passing when running the same spec in isolation and so forth. Though, I see a huge value if this was more stable, currently it's a full time job trying to get e2e pass and most of failures are usually false.  Would love to make it part of CI at some point.

**Comment by tymfear** [comment]
@demisx You cannot use Jasmine 3 with Protractor version lower than 6, since Protractor itself used jasminewd adapter, which uses Jasmine 2.x. Do you use async/await and promise manager simultaneously? This is an often reason of such behavior. I got 1300 tests running on CI and it goes well.
Also, pay closer attention to configuring headless chrome. But the better approach would be using dockerized selenium. I recommend to use [Selenoid](https://github.com/aerokube/selenoid).

**Comment by demisx** [comment]
@tymfear Thank you very much for the info and your recommendations! Please see my responses below.

> You cannot use Jasmine 3 with Protractor version lower than 6, since Protractor itself used jasminewd adapter, which uses Jasmine 2.x

Thank you. Didn't know about this.  I am going to try to upgrade to v6 or downgrade Jasmine to 2.x to see if it makes a difference. 

> Do you use async/await and promise manager simultaneously? This is an often reason of such behavior

No, we use `async/await` with `SELENIUM_PROMISE_MANAGER: false`

> I got 1300 tests running on CI and it goes well.

This is vey motivating to hear. Thank you for giving me another hope. We were about to give up on this dream. 

> Also, pay closer attention to configuring headless chrome. But the better approach would be using dockerized selenium. I recommend to use Selenoid.

Excellent. I'll give Selenoid a try! Feel like I may just become a born again Protractorist.

**Comment by yassinchaibi** [comment]
Hello @tymfear,
we tried to change the jasmine version but we discovered that protractor 5.4.2 installs the jasmine 2.8.0 by default, and we were unable to change the version of it, is there any solution for this?
Thank you

**Comment by tymfear** [comment]
@yassinchaibi no, there is no. Only upgrade to Protractor 6, where jasminewd was removed and you are free to use any version of Jasmine

**Comment by yassinchaibi** [comment]
@tymfear we are very thankfull for your replay.
the think is we are forced to use Protractor v5.4.2 due to project restrection :/ .

**Comment by Kurt29** [comment]
@tymfear
Is protractor 6 usable right now? I thought it has a lot of problems and therefor postponed to fix it?

**Comment by MojoCC** [comment]
@Kurt29 @tymfear 
I would to know too. I have a lot of problems with Protractor 5...

**Comment by ssudhanshu** [comment]
> @demisx You cannot use Jasmine 3 with Protractor version lower than 6, since Protractor itself used jasminewd adapter, which uses Jasmine 2.x. Do you use async/await and promise manager simultaneously? This is an often reason of such behavior. I got 1300 tests running on CI and it goes well.
> Also, pay closer attention to configuring headless chrome. But the better approach would be using dockerized selenium. I recommend to use [Selenoid](https://github.com/aerokube/selenoid).

Hi @yassinchaibi ,
I am also using Protractor to run my web-based tests on CI with around 1000 tests running daily. It was all going well until all my tests started to fail. The reason: the web-based app depend upon login, and due to a minor issue with app, login failed, followed by that all my tests failed and it took 8hrs to complete! This was a big mess. I started to wonder how can I avoid such failures. I checked and found fail-fast mode which did no good because it would stop execution on the first failure. I want something that stops the execution as soon as login fails. Is there a way to work this out?
Tests are run in docker in headless mode.
Thanks.

**Comment by demisx** [comment]
@ssudhanshu You are probably better off asking this question on SO (stackoverflow.com) and it's off topic for this read.

**Comment by CrispusDH** [comment]
You can run script with login before tests and if it fails you can not start tests. Write script, run it using package.json `preNameOfTestScript`.

**Comment by ingriddbritok3** [comment]
> @demisx You cannot use Jasmine 3 with Protractor version lower than 6, since Protractor itself used jasminewd adapter, which uses Jasmine 2.x. Do you use async/await and promise manager simultaneously? This is an often reason of such behavior. I got 1300 tests running on CI and it goes well.
> Also, pay closer attention to configuring headless chrome. But the better approach would be using dockerized selenium. I recommend to use [Selenoid](https://github.com/aerokube/selenoid).

Hi @tymfear 
I am using Jasmine 6.0
Chome driver 80.0
Angular version of the site 6.1.7
Node version v12.14.1.
On my config file I have directConnect: true,
On my spec file I am using async/await.
The behavior of the tests is unstable. On the first run it passes, on the second run ,it doesnt find the element. I also tried to modified the config file adding SELENIUM_PROMISE_MANAGER: false, but then the application could not be started. Is there anything I am missing?
Tks!

**Comment by demisx** [comment]
@ingriddbritok3 We've switched to Cypress, since Protractor was giving us too many headaches. No  stability issues so far.

**Comment by SchnWalter** [comment]
> I also tried to modified the config file adding SELENIUM_PROMISE_MANAGER: false, but then the application could not be started. Is there anything I am missing?

@ingriddbritok3, did you update your tests to deal with all the async operations?

**Comment by ingriddbritok3** [comment]
@SchnWalter Yes I also tried this, but this does not resolve the instability of Protractor

**Comment by SchnWalter** [comment]
> In the console, the name of the test being executed is only displayed on the console after the test passes or fails.

@Inframatic, what SpecReporter are you using? This is the job of the spec reporter, it has nothing to do with protractor. For example, if you are using the [`jasmine-spec-reporter`](https://github.com/bcaudan/jasmine-spec-reporter/blob/v5.0.1/src/spec-reporter.ts#L91-L119), you can see in the code that the spec name is not displayed in `SpecReporter.specStarted()`, but only in `SpecReporter.specDone()`, so you might wanna open an issue on their repository to provide an improved `DisplayProcessor` that can deal with this requirement.
